engine/common/follower/core.go

```diff
@@ -0,0 +1,317 @@
+// Here we process events that need to be sequentially ordered(processing certified blocks and new finalized blocks).
+func (c *Core) processCoreSeqEvents(ctx irrecoverable.SignalerContext, ready component.ReadyFunc) {
+	ready()
+
+	doneSignal := ctx.Done()
+	for {
+		select {
+		case <-doneSignal:
+			return
+		case finalized := <-c.finalizedBlocksChan:
+			err := c.processFinalizedBlock(ctx, finalized) // no errors expected during normal operations
+			if err != nil {
+				ctx.Throw(err)
+			}
+		case blocks := <-c.certifiedBlocksChan:
+			err := c.processCertifiedBlocks(ctx, blocks) // no errors expected during normal operations
+			if err != nil {
+				ctx.Throw(err)
+			}
+		}
+	}
+}
+
+// OnFinalizedBlock updates local state of pendingCache tree using received finalized block and queues finalized block
+// to be processed by internal goroutine.
+// This function is safe to use in concurrent environment.
+func (c *Core) OnFinalizedBlock(final *flow.Header) {
+	c.pendingCache.PruneUpToView(final.View)
+
+	// in-case we have already stopped our worker we use a select statement to avoid
+	// blocking since there is no active consumer for this channel
+	select {
+	case c.finalizedBlocksChan <- final:
+	case <-c.ComponentManager.ShutdownSignal():
+	}
+}
+
+// processCertifiedBlocks process a batch of certified blocks by adding them to the tree of pending blocks.
+// As soon as tree returns a range of connected and certified blocks they will be added to the protocol state.
+// Is NOT concurrency safe, has to be used by internal goroutine.
+// No errors expected during normal operations.
+func (c *Core) processCertifiedBlocks(ctx context.Context, blocks CertifiedBlocks) error {
+	span, ctx := c.tracer.StartSpanFromContext(ctx, trace.FollowerProcessCertifiedBlocks)
+	defer span.End()
+
+	connectedBlocks, err := c.pendingTree.AddBlocks(blocks)
+	if err != nil {
+		return fmt.Errorf("could not process batch of certified blocks: %w", err)
+	}
+	err = c.extendCertifiedBlocks(ctx, connectedBlocks)
+	if err != nil {
+		return fmt.Errorf("could not extend protocol state: %w", err)
+	}
+	return nil
+}
+
+// extendCertifiedBlocks processes a connected range of certified blocks by applying them to protocol state.
+// As result of this operation we might extend protocol state.
+// Is NOT concurrency safe, has to be used by internal goroutine.
+// No errors expected during normal operations.
+func (c *Core) extendCertifiedBlocks(parentCtx context.Context, connectedBlocks CertifiedBlocks) error {
+	span, parentCtx := c.tracer.StartSpanFromContext(parentCtx, trace.FollowerExtendCertifiedBlocks)
+	defer span.End()
+
+	for _, certifiedBlock := range connectedBlocks {
+		span, ctx := c.tracer.StartBlockSpan(parentCtx, certifiedBlock.ID(), trace.FollowerExtendCertified)
+		err := c.state.ExtendCertified(ctx, certifiedBlock.Block, certifiedBlock.QC)
+		span.End()
+		if err != nil {
+			if state.IsOutdatedExtensionError(err) {
+				continue
+			}
```

Comment:
## ⚠️ 

I am worried that the potential of the protocol state returning an `OutdatedExtension` error here can create the following edge case:
* Imagine the following fork
   ```
    <- A <-B <-C <-D <- E <- F <- G
           ^- X <- Y <- Z
   ``` 
   where A is finalized. The blocks `B, C, D, E, F, G` have already been ingested and forwarded to the consensus follower loop.
* Assume that the consensus follower loop has not yet processed `B, .., G`, but when it does, it finalizes `B, C`
* Now block `X` comes in and, and `X,Y` are returned as a `connectedBlocks`
* Race-condition edge case:
   - follower loop processes  `B, .., G` and finalizes up to B
   - we now call line 262: we call `state.ExtendCertified( X )`   which returns `OutdatedExtensionError`
   - then, we process block `Y`, whose parent is not known by the protocol state, which causes an exception 💀 

Suggestion:
* With the new logic, I don't think the protocol state should decide which forks are orphaned. Instead, it should just store whatever it is asked to store (still checking that the parent is known)
* All the logic of tracking orphaned blocks is now in the pending tree.

Reply:
ok, that makes, I will get rid of it